### PR TITLE
Generate a single hash

### LINKED ISSUE #1330: Provide a way to limit to a single hash per requirement when generating hashes
#### What's the problem this feature will solve?

Currently, when the `--generate-hashes` option is passed to `pip-compile`, `pip-compile` will include potentially many hashes per requirement, even if only one is needed / desired.

#### Describe the solution you'd like

This feature request is for `pip-compile` to expose a command-line option that would limit the output to including only a single hash, namely the hash of what would be installed in the environment in which `pip-compile` is being run.

This option would be useful in situations where the exact deployment target is known (e.g. when using containers). For one, the requirements files generated by `pip-compile` would be shorter and easier to review since they wouldn't include extraneous info. Secondly, I believe this would provide greater determinism / reproducibility. For example, currently, if a release that was previously being installed from a requirements file generated by `pip-compile` was [yanked](https://pypi.org/help/#yanked) from PyPI, then the result of what would be installed from that requirements file would change, even though the requirements file didn't change (because the requirements file currently includes more hashes than what was originally installed).

### PR DESCRIPTION
**Pull request by plannigan**
**Changelog-friendly one-liner**: Option to restrict generated hashes to the single best matching file

Closes #1330 

##### Contributor checklist

- [x] Provided the tests for the changes.
- [x] Gave a clear one-line description in the PR (that the maintainers can add to CHANGELOG.md on release).
- [x] Assign the PR to an existing or new milestone for the target version (following [Semantic Versioning](https://blog.versioneye.com/2014/01/16/semantic-versioning/)).

### CODE REVIEW

**Comment by plannigan** [comment]
The `mypy` failures in the pre-commit.ci look like they are legitimate, but not related to my changes. I am getting the same errors when running `checkqa` locally on the default branch.

**Comment by plannigan** [comment]
Pulled in the latest changes to the default branch and now `mypy` is passing. The rest of the CI is waiting for a maintainer to approve it to run.

**Comment by plannigan** [comment]
I switched to the compatibility implementation of `nullcontext()` which will resolve the previous failure on python 3.6.

**Review by ssbarnea — APPROVED** [review]
While I do not see myself using it, it does still LGTM, especially as the number of hashes
can become quite large for some packages.

Still, I hope others will review the change better and note if there are reasons for not doing this change.

**Comment by atugushev** [comment]
Can we postpone this to 6.3.0? There is no rush and i'd like to take a look at this later.

**Comment by plannigan** [comment]
Pulled in master and added a test case for `single_hash` the FakeRepository. That should make CodeCov happy.

Looks like 6.2.0 was released earlier this week, so feel free to merge when you are ready.

**Comment by ssbarnea** [comment]
Hint for @plannigan : to avoid waiting to workflow approval you could try to make a very small contribution first, like fixing a typo (i bet there is one to fix). Once you already contributed successfully workflow approval is no longer needed. Another option would to join https://jazzband.co/ which is automated if I remember well.

**Review by atugushev — COMMENTED** [review]
I've tried locally, but the output still has many hashes. Check this out:

```ini
# requirements.in
psycopg2
```

```console
$ pip-compile --generate-hashes --single-hash
#
# This file is autogenerated by pip-compile with python 3.8
# To update, run:
#
#    pip-compile --generate-hashes --single-hash
#
psycopg2==2.9.1 \
    --hash=sha256:079d97fc22de90da1d370c90583659a9f9a6ee4007355f5825e5f1c70dffc1fa \
    --hash=sha256:2087013c159a73e09713294a44d0c8008204d06326006b7f652bef5ace66eebb \
    --hash=sha256:2c992196719fadda59f72d44603ee1a2fdcc67de097eea38d41c7ad9ad246e62 \
    --hash=sha256:7640e1e4d72444ef012e275e7b53204d7fab341fb22bc76057ede22fe6860b25 \
    --hash=sha256:7f91312f065df517187134cce8e395ab37f5b601a42446bdc0f0d51773621854 \
    --hash=sha256:830c8e8dddab6b6716a4bf73a09910c7954a92f40cf1d1e702fb93c8a919cc56 \
    --hash=sha256:89409d369f4882c47f7ea20c42c5046879ce22c1e4ea20ef3b00a4dfc0a7f188 \
    --hash=sha256:bf35a25f1aaa8a3781195595577fcbb59934856ee46b4f252f56ad12b8043bcf \
    --hash=sha256:de5303a6f1d0a7a34b9d40e4d3bef684ccc44a49bbe3eb85e3c0bffb4a131b7c
    # via -r requirements.in
```

**Comment by plannigan** [comment]
I am not able to reproduce that result (see my results below). @atugushev Can you provide some more information on your setup and how you got the above output? 

```
(.venv) C:\Users\plannigan\code\pip-tools>git log -1
commit 876fcb2a6938da53c9590f9dc7d44f1564d9d62d (HEAD -> single-hash, origin/single-hash)
Author: Patrick Lannigan <p.lannigan@gmail.com>
Date:   Fri Jun 25 07:53:50 2021 -0400

    Add test case for fake repository with single hash

(.venv) C:\Users\plannigan\code\pip-tools>git status
On branch single-hash
Your branch is up to date with 'origin/single-hash'.

nothing to commit, working tree clean

(.venv) C:\Users\plannigan\code\pip-tools>python --version
Python 3.8.6

(.venv) C:\Users\plannigan\code\pip-tools>pip list
Package    Version              Location
---------- -------------------- ------------------------------
click      8.0.1
colorama   0.4.4
pep517     0.10.0
pip        21.1.3
pip-tools  6.2.1.dev14+g876fcb2 c:\users\plannigan\code\pip-tools
setuptools 49.2.1
toml       0.10.2
wheel      0.36.2

(.venv) C:\Users\plannigan\code\pip-tools>type requirements.in
psycopg2

(.venv) C:\Users\plannigan\code\pip-tools> pip-compile --generate-hashes --single-hash
#
# This file is autogenerated by pip-compile with python 3.8
# To update, run:
#
#    pip-compile --generate-hashes --single-hash
#
psycopg2==2.9.1 \
    --hash=sha256:2c992196719fadda59f72d44603ee1a2fdcc67de097eea38d41c7ad9ad246e62
    # via -r requirements.in
```

**Comment by atugushev** [comment]
It looks like there is an issue with existing hashes in `requirements.txt`. How to reproduce:

1. Generate hashes first:
```console
$ pip-compile --generate-hashes
#
# This file is autogenerated by pip-compile with python 3.8
# To update, run:
#
#    pip-compile --generate-hashes
#
psycopg2==2.9.1 \
    --hash=sha256:079d97fc22de90da1d370c90583659a9f9a6ee4007355f5825e5f1c70dffc1fa \
    --hash=sha256:2087013c159a73e09713294a44d0c8008204d06326006b7f652bef5ace66eebb \
    --hash=sha256:2c992196719fadda59f72d44603ee1a2fdcc67de097eea38d41c7ad9ad246e62 \
    --hash=sha256:7640e1e4d72444ef012e275e7b53204d7fab341fb22bc76057ede22fe6860b25 \
    --hash=sha256:7f91312f065df517187134cce8e395ab37f5b601a42446bdc0f0d51773621854 \
    --hash=sha256:830c8e8dddab6b6716a4bf73a09910c7954a92f40cf1d1e702fb93c8a919cc56 \
    --hash=sha256:89409d369f4882c47f7ea20c42c5046879ce22c1e4ea20ef3b00a4dfc0a7f188 \
    --hash=sha256:bf35a25f1aaa8a3781195595577fcbb59934856ee46b4f252f56ad12b8043bcf \
    --hash=sha256:de5303a6f1d0a7a34b9d40e4d3bef684ccc44a49bbe3eb85e3c0bffb4a131b7c
    # via -r requirements.in
```

2. Then run `pip-compile` again with the single hash option:
```console
$ pip-compile --generate-hashes --single-hash
#
# This file is autogenerated by pip-compile with python 3.8
# To update, run:
#
#    pip-compile --generate-hashes --single-hash
#
psycopg2==2.9.1 \
    --hash=sha256:079d97fc22de90da1d370c90583659a9f9a6ee4007355f5825e5f1c70dffc1fa \
    --hash=sha256:2087013c159a73e09713294a44d0c8008204d06326006b7f652bef5ace66eebb \
    --hash=sha256:2c992196719fadda59f72d44603ee1a2fdcc67de097eea38d41c7ad9ad246e62 \
    --hash=sha256:7640e1e4d72444ef012e275e7b53204d7fab341fb22bc76057ede22fe6860b25 \
    --hash=sha256:7f91312f065df517187134cce8e395ab37f5b601a42446bdc0f0d51773621854 \
    --hash=sha256:830c8e8dddab6b6716a4bf73a09910c7954a92f40cf1d1e702fb93c8a919cc56 \
    --hash=sha256:89409d369f4882c47f7ea20c42c5046879ce22c1e4ea20ef3b00a4dfc0a7f188 \
    --hash=sha256:bf35a25f1aaa8a3781195595577fcbb59934856ee46b4f252f56ad12b8043bcf \
    --hash=sha256:de5303a6f1d0a7a34b9d40e4d3bef684ccc44a49bbe3eb85e3c0bffb4a131b7c
    # via -r requirements.in
```

---

Note that `--no-reuse-hashes` helps to generate single hash output:
```console
$ pip-compile --generate-hashes --no-reuse-hashes --single-hash
#
# This file is autogenerated by pip-compile with python 3.8
# To update, run:
#
#    pip-compile --generate-hashes --no-reuse-hashes --single-hash
#
psycopg2==2.9.1 \
    --hash=sha256:de5303a6f1d0a7a34b9d40e4d3bef684ccc44a49bbe3eb85e3c0bffb4a131b7c
    # via -r requirements.in
```

**Comment by plannigan** [comment]
Ok, that makes sense. The local repository checks the cache before calling `get_hashes()` for the PyPI repository. I'll have some time to circle back on this in a few days.

**Comment by plannigan** [comment]
Thanks for tracking down the hash reuse issue. The fix was a simple check to see if there were more than one existing hash.

**Review by AndydeCleyre — CHANGES_REQUESTED** [review]
Small English changes in the readme, please:

- "give" to "given"
- "match" to "matching"

Thanks!

File: piptools/repositories/base.py
Comment (AndydeCleyre) [inline piptools/repositories/base.py:43]: Small typo: please change "included" to "include"

File: piptools/repositories/pypi.py
Comment (AndydeCleyre) [inline piptools/repositories/pypi.py:310]: Please change this one, too: "included" to "include" or "includes"

**Comment by plannigan** [comment]
Fixed the grammar in a few more locations in docstrings. Should be good to go now.

**Comment by ssbarnea** [comment]
Closing this PR as it seems outdated. Please try to reopen it once these problems are sorted. Small PRs that are easy to review are likely to be merged quickly as they impose less risks.

**Pull request closed without merging**
